src/client/scripts/game/chess/premove.js

```diff
@@ -0,0 +1,248 @@
+            } else {
+                pieceTranslation.startCoords = premove.endCoords;
+            }
+        }
+
+        //Un-highlight the square the piece moved from
+        removeSquareHighlight(premove.startCoords);
+    }
+    
+    /**
+     * Move the visual position of the piece if premoves are shown.
+     * @param {Piece} piece - The piece to premove.
+     * @param {number[] | null} newCoords - The coordinates to move the piece to. *null* if the piece was captured.
+     */
+    function premovePiece(piece, newCoords) {
+        if(newCoords) {
+            
+            let capturedPiece = getPieceAtCoords(newCoords);
+            if (capturedPiece) premovePiece(capturedPiece, null);
+
+            //Update the visual position of the piece.
+            piecesmodel.movebufferdata(game.getGamefile(), piece, newCoords);
+        } else {
+            //The piece was captured. Remove it.
+            piecesmodel.deletebufferdata(game.getGamefile(), piece);
+        }
+
+        piece.premovedCoords = newCoords;
+        movedPieces.push(piece);
+    }
+
+    /** Remove premove highlight from a square.
+     * @pram {number[]} coords - The coordinates of the square to un-highlight
+     */
+    function removeSquareHighlight(coords) {
+        let highlighedSquareIndex = highlightedSquares.indexOf(coords)
+        if (highlighedSquareIndex < 0)
+            return console.error("Cannot remove highlight as it was never added.");
+        highlightedSquares.splice(highlighedSquareIndex, 1);
+    }
+
+    /** Adds a premove to the queue.
+     * @param {Piece} piece - the piece that was moved
+     * @param {Move} move - the move the piece made
+    */
+    function makePremove(piece, move) {
+        if (!premovesEnabled)
+            return;
+        if (main.devBuild) console.log("A premove was made.");
+        premoves.push(move);
+
+        let trimmedType = math.trimWorBFromType(piece.type);
+        let specialMoveMade;
+        let gamefile = game.getGamefile();
+        if(gamefile.specialMoves[trimmedType]) 
+            specialMoveMade = game.getGamefile().specialMoves[trimmedType](gamefile, piece, move, { isPremove:true });
+
+        if (!specialMoveMade) premovePiece(piece, move.endCoords);
+
+        highlightedSquares.push(move.endCoords);
+    }
+
+    /** Sends all premoved pieces back to their original positions then clears the queue of premoves. */
+    function clearPremoves()
+    {
+        hidePremoves();
+        premoves = [];
+        movedPieces = [];
+        highlightedSquares = [];
+    }
+
+    /** Displays premoved pieces in their new positions. */
+    function showPremoves() {
+        if(!premoves) return;
+        let gamefile = game.getGamefile();
+        for (let movedPiece of Object.values(movedPieces)) {
+            if(movedPiece.premovedCoords)
+                piecesmodel.movebufferdata(gamefile, movedPiece, movedPiece.premovedCoords);
```

Comment:
Hmm. This seems a bit weird for premove.js to modify the mesh data of the gamefile. It may be better to just use `movepiece.makeMove()` instead. That handles everything for you from modifying the mesh data to animating the piece to sound effects to updating the organized lines. I imagine not using would be the source of many bugs.

Here's how it would work. When we make a premove, we call `makeMove()`. When we receive our opponent's move, at that time we use`movepiece.rewindGameToIndex()` to undo all our premoves, then we make our opponents move, then we check if the first premove in our list is legal, if so we make it, and re-do the the remainder of the premoves.

Reply:
Thank you. This has now been implemented.